ipie/addons/thermal/estimators/energy.py

```diff
@@ -0,0 +1,89 @@
+# Copyright 2022 The ipie Developers. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Authors: Fionn Malone <fmalone@google.com>
+#          Joonho Lee
+#
+
+from typing import Union
+
+from ipie.utils.backend import arraylib as xp
+from ipie.hamiltonians.generic import GenericComplexChol, GenericRealChol
+from ipie.estimators.estimator_base import EstimatorBase
+
+from ipie.addons.thermal.walkers.uhf_walkers import UHFThermalWalkers
+from ipie.addons.thermal.estimators.thermal import one_rdm_from_G
+from ipie.addons.thermal.estimators.generic import local_energy_generic_cholesky
+
+
+def local_energy(
+        hamiltonian: Union[GenericRealChol, GenericComplexChol], 
+        walkers: UHFThermalWalkers):
+    energies = xp.zeros((walkers.nwalkers, 3), dtype=xp.complex128)
+    
+    for iw in range(walkers.nwalkers):
+        # Want the full Green's function when calculating observables.
+        walkers.calc_greens_function(iw, slice_ix=walkers.stack[iw].nslice)
+        P = one_rdm_from_G(xp.array([walkers.Ga[iw], walkers.Gb[iw]]))
+        energy = local_energy_generic_cholesky(hamiltonian, P)
+        energies[iw] = energy
+
+    return energies
+
+
+class ThermalEnergyEstimator(EstimatorBase):
```

Comment:
I think this is essentially identical to EnergyEstimator right? You could save some space and do:

``` python
class ThermalEnergyEstimator(EnergyEstimator):
    # Only overwrite the single method compute_estimator to have the right import.
```